### ISSUE TITLE
Support for 403

### ISSUE THREAD

**Issue by jayadeba** [graphql/graphql-over-http#149]
Hi,

I see that the proposed spec is supporting most HTTP status codes. Does this group has any opinion on supporting the following;

1-Add support for 403 (Can generally be thrown when the caller doesn't have permissions to execute an operation ( a query or mutation) or any related authorization error)

2-Use the equivalent HTTP code for "Request Entity Too Large" 413 instead of 400 (https://github.com/graphql/express-graphql/issues/702)

3- Have some sort of statement on whether an API provider can use any error representation for 4XX and 5XX errors. See https://tools.ietf.org/html/draft-nottingham-http-problem-03 

4- Have more Concrete Guidance around 4XX (400 in particular) as can extension to the statements that are made to the status codes section , with clear examples. For example an API Provider MAY throw a 400 when if request validation fails 

5- Provide Concrete guidance around 5XX. Most of 5XX are recoverable errors so retries from the clients may result in successful processing of the request (In general all intermediaries are configured to do a retry) so it can left up to the API provider to throw a 500 from downstream failure.

Happy to discuss these and contribute back.

Thanks,
Jayadeba

**Comment by benjie** [comment]
I've added this to the 1.0 Release not because we'll address all these points, but because I want to make sure we at least consider them, even if we decide to take no action at this time. We'll probably move this to a later release (or close it) depending on the outcome of these considerations. Thanks for the input @jayadeba :+1:

**Comment by martinbonnin** [comment]
Hey @jayadeba. Thanks for sending that list. That's a lot of things! Few feedbacks:

> Providing guidance in the specification to use appropriate HTTP codes, particularly around the HTTP error codes 4XX/5XX to describe the outcome, without being prescriptive about which error code should be used when, would be helpful.

There is now a whole section about it there: https://graphql.github.io/graphql-over-http/draft/#sel-DANNDTAAEVDEQq-Z

The general rule is:
* if there is `data`, return `2xx`
* if there is no `data`, return `4xx` or `5xx` as appropriate. The spec gives some recommendations but does not prescribe the exact status codes (just that they are `4xx` or `5xx`).

> Describing how to perform Authentication Authorization for GraphQL requests over HTTP 

That sounds interesting. I'm not aware of anything yet. Could maybe be a best practice?

> Providing some guidance/making statements around File handling 

Did you mean file upload? There is a best practice there: https://graphql.org/learn/file-uploads/

> Providing some guidance around idempotent mutation operations over HTTP

I'm confused. Mutations are typically not idempotent (contrary to queries). Can you ellaborate a bit more?

> covering all HTTP error codes would be helpful for the Implementor 

There is a good bunch covered at https://graphql.github.io/graphql-over-http/draft/#sel-DANNDTAAEVDEQq-Z

> coming up with directives to describe HTTP protocol specific metadata in the schema would be helpful

See https://github.com/graphql/graphql-over-http/issues/150. I'm going to recommend to do this in [GAPs](https://gaps.graphql.org/), which have been done exactly for that purpose.

I know it's been a long time. If you're still around, please feel free to comment here or even open more specific issues for some of the points above, either in this repo or in the GAPs repo.

In the absence of response from you, I'll close this issue as non-actionable.

**Comment by benjie** [comment]
Mutations are idempotent if you can issue them multiple times and the outcome is the same. A patch (update) mutation is typically idempotent unless you use updatedAt timestamps or issue events on update, a delete mutation is idempotent if it gives the same result whether or not the record existed before the delete (assuming no events). Other mutation types may or may not be idempotent.

**Comment by martinbonnin** [comment]
> A patch (update) mutation is typically idempotent unless you use updatedAt timestamps or issue events on update

Did we want to investigate caching those mutations? Is there prior art in HTTP?

**Comment by benjie** [comment]
> Did we want to investigate caching those mutations? 

No, I don't think so. The point of idempotency for mutations is if the connection is interrupted (e.g. if you're doing it from a mobile connection) and you don't receive confirmation, you can safely issue the same mutation again without worrying that you'll change something. For example repeating `mutation {postMessage(channelId:123,message:"hi"){...}}` could result in the same message being posted multiple times; whereas `mutation postMessage(channelId:123, message:"hi", nonce:123456789){...}}` could use the nonce to ensure the mutation only happens once - next time, a different nonce would need to be supplied to perform the next mutation.

**Comment by martinbonnin** [comment]
Makes sense 👍
